File: leetcode/python/429_N-ary_tree_level_order_traversal.py
from collections import deque


class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children

class Solution:
    def levelOrder(self, root: Node | None) -> list[list[int]]:
        if not root:
            return []
        res: list[list[int]] = [[]]
        q = deque([root, None])
        while q:
            node = q.popleft()
            if not node:
                if q:
                    res.append([])
                    q.append(None)
                continue
            res[-1].append(node.val)
            for child in node.children:
                q.append(child)
        return res

File: leetcode/c++/429_N-ary_tree_level_order_traversal.cpp
#include <vector>
#include <queue>

using namespace std;


// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val) {
        val = _val;
    }

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};

class Solution {
public:
    vector<vector<int>> levelOrder(Node* root) {
        vector<vector<int>> res = {};
        queue<Node*> q({root});
        while (!q.empty()) {
            res.push_back({});
            int size = q.size();
            for (int i = 0; i < size; ++i) {
                auto node = q.front();
                q.pop();
                res.back().push_back(node->val);
                for (auto child : node->children) {
                    q.push(child);
                }
            }
        }
        return res;
    }
};